src/pc/macros.c

```diff
@@ -0,0 +1,98 @@
+#include <game.h>
+#ifdef _MSC_VER
+#define SDL_MAIN_HANDLED
+#include <SDL.h>
+#else
+#include <SDL2/SDL.h>
+#endif
+#include "macros.h"
+
+int press_pos = 0;
+int press_spell = 0;
+int next_time = 0;
+int spell_num = 0;
+int spell_dir = 0;
+
+#define STOP -1
+
+struct SpellInput {
+    int pad;
+    int frames;
+};
+
+struct SpellInput tetra_spirit[] = {
+    {PAD_UP, 33},   {PAD_UP + PAD_RIGHT, 1},
+    {PAD_RIGHT, 1}, {PAD_DOWN + PAD_RIGHT, 1},
+    {PAD_DOWN, 1},  {PAD_SQUARE, 1},
+    {STOP, 0}};
+
+struct SpellInput dark_metamorphosis[] = {
+    {PAD_LEFT, 1},  {PAD_UP + PAD_LEFT, 1},
+    {PAD_UP, 1},    {PAD_UP + PAD_RIGHT, 1},
+    {PAD_RIGHT, 1}, {PAD_SQUARE, 1},
+    {STOP, 0}};
+
+struct SpellInput* spells_right[] = {&tetra_spirit, &dark_metamorphosis};
+
+struct SpellInput tetra_spirit_left[] = {
+    {PAD_UP, 33},  {PAD_UP + PAD_LEFT, 1},
+    {PAD_LEFT, 1}, {PAD_DOWN + PAD_LEFT, 1},
+    {PAD_DOWN, 1}, {PAD_SQUARE, 1},
+    {STOP, 0}};
+
+struct SpellInput dark_metamorphosis_left[] = {
+    {PAD_RIGHT, 1}, {PAD_UP + PAD_RIGHT, 1},
+    {PAD_UP, 1},    {PAD_UP + PAD_LEFT, 1},
+    {PAD_LEFT, 1},  {PAD_SQUARE, 1},
+    {STOP, 0}};
+
+struct SpellInput* spells_left[] = {
+    &tetra_spirit_left, &dark_metamorphosis_left};
+
+struct SpellInput** spell[] = {&spells_right, &spells_left};
+
+enum Spells { TETRA_SPIRIT, DARK_METAMORPHOSIS };
+
+u8 previousKeyStates[SDL_NUM_SCANCODES] = {0};
+
+void check_macro(int scanCode, enum Spells which, const u8* keyb) {
+    if (keyb[scanCode] && !previousKeyStates[scanCode]) {
+        press_spell = 1;
+        press_pos = 0;
+        next_time = 0;
+        spell_num = which;
+        spell_dir = PLAYER.facingLeft;
+    }
+}
+
+void handle_macros(const u8* keyb, u_long* pressed) {
```

Comment:
I wonder if we can leverage `DemoUpdate(void)` and `g_DemoMode` to pass macros? nothing that needs to be done now, this is good as it is

Reply:
Looking at the usages of `g_DemoMode`, there's a bunch of entanglements that change the state of the game so I'm thinking it's cleaner to not use it.